File: LSP+ISP/Spoilers/MarsRoverKata/Projectile.cs
using System.Collections.Generic;
using System.Linq;

namespace MarsRoverKata
{
    public abstract class Projectile
    {
        protected Projectile(Mars mars)
        {
            Mars = mars;
        }

        private Mars Mars { get; set; }
        private static readonly Dictionary<Direction, Point> PositionalAdjustments = new Dictionary<Direction, Point>
        { 
            { Direction.North, new Point(0, 1) },
            { Direction.South, new Point(0, -1) },
            { Direction.East, new Point(1, 0) },
            { Direction.West, new Point(-1, 0) }
        };

        protected abstract int MaxRange { get; }

        public void Launch(Direction facing, Point location)
        {
            bool collidedWithTarget = false;
            int moveIndex = 0;
            Point target = location;
            while (!collidedWithTarget && moveIndex < MaxRange)
            {
                Point desiredPosition = CreateDesiredPosition(1, facing, target);
                target = CalculateProjectileFinalPosition(target, desiredPosition);
                collidedWithTarget = IsCollisionDetected(target);
                moveIndex++;
            }
            HitTarget(target);
        }

        private static Point CreateDesiredPosition(int adjustmentFactor, Direction facing, Point location)
        {
            var adjustment = PositionalAdjustments[facing] * adjustmentFactor;
            return location + adjustment;
        }

        private Point CalculateProjectileFinalPosition(Point from, Point desired)
        {
            Point newDestination = desired;
            newDestination = CalculatePositionY(desired, newDestination);
            newDestination = CalculatePositionX(desired, newDestination);

            return newDestination;
        }

        protected abstract bool IsCollisionDetected(Point desired);

        protected Point CalculatePositionX(Point desired, Point newDestination)
        {
            if (desired.X > Mars.Bounds.Width)
            {
                newDestination = new Point(0, desired.Y);
            }
            if (desired.X < 0)
            {
                newDestination = new Point(Mars.Bounds.Width, desired.Y);
            }
            return newDestination;
        }

        protected Point CalculatePositionY(Point desired, Point newDestination)
        {
            if (desired.Y > Mars.Bounds.Height)
            {
                newDestination = new Point(desired.X, 0);
            }
            if (desired.Y < 0)
            {
                newDestination = new Point(desired.X, Mars.Bounds.Height);
            }
            return newDestination;
        }

        protected Obstacle FindObstacle(Point point)
        {
            return Mars.Obstacles.SingleOrDefault(x => x.Location.Equals(point));
        }

        private void HitTarget(Point point)
        {
            var obstacle = FindObstacle(point);

            if (obstacle != null && obstacle.IsDestructable)
            {
                DestroyObstacle(obstacle);
            }
            else if (obstacle == null)
            {
                CreateObstacle(point);
            }
        }

        protected virtual void DestroyObstacle(Obstacle obstacle)
        {
            Mars.RemoveObstacle(obstacle);
        }

        protected virtual void CreateObstacle(Point point)
        {
            Obstacle obstacle = new Crater(point);
            Mars.AddObstacle(obstacle);
        }
    }
}

File: LSP+ISP/Java/MarsRoverKata/src/main/java/MarsRoverKata/Projectile.java
package MarsRoverKata;

import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.common.collect.Iterables;

import java.util.HashMap;
import java.util.Map;

public abstract class Projectile {

    private static Map<Direction, Point> positionalAdjustments = new HashMap<Direction, Point>() {{
        put(Direction.North, new Point(0, 1));
        put(Direction.South, new Point(0, -1));
        put(Direction.East, new Point(1, 0));
        put(Direction.West, new Point(-1, 0));
    }};

    protected Mars mars;
    protected int maxRange;

    protected Projectile(Mars mars) {
        this.mars = mars;
    }

    protected abstract int getMaxRange();

    public void launch(Direction facing, Point location) {
        boolean collidedWithTarget = false;
        int moveIndex = 0;
        Point target = location;
        while (!collidedWithTarget && moveIndex < getMaxRange()) {
            Point desiredPosition = createDesiredPosition(1, facing, target);
            target = calculateProjectileFinalPosition(target, desiredPosition);
            collidedWithTarget = isCollisionDetected(target);
            moveIndex++;
        }
        hitTarget(target);
    }

    private static Point createDesiredPosition(int adjustmentFactor, Direction facing, Point location) {
        Point adjustment = Point.multiply(positionalAdjustments.get(facing), (adjustmentFactor));
        return Point.add(location, adjustment);
    }

    private Point calculateProjectileFinalPosition(Point from, Point desired) {
        Point newDestination = desired;
        newDestination = calculatePositionY(desired, newDestination);
        newDestination = calculatePositionX(desired, newDestination);

        return newDestination;
    }

    protected abstract boolean isCollisionDetected(Point desired);

    protected Point calculatePositionX(Point desired, Point newDestination) {
        if (desired.getX() > mars.getBounds().getWidth()) {
            newDestination = new Point(0, desired.getY());
        }
        if (desired.getX() < 0) {
            newDestination = new Point(mars.getBounds().getWidth(), desired.getY());
        }
        return newDestination;
    }

    protected Point calculatePositionY(Point desired, Point newDestination) {
        if (desired.getY() > mars.getBounds().getHeight()) {
            newDestination = new Point(desired.getX(), 0);
        }
        if (desired.getY() < 0) {
            newDestination = new Point(desired.getX(), mars.getBounds().getHeight());
        }
        return newDestination;
    }

    protected IObstacle findObstacle(final Point point) {
        return Iterables.getOnlyElement(Collections2.filter(mars.getObstacles(), new Predicate<IObstacle>() {
            @Override
            public boolean apply(IObstacle input) {
                return input.getLocation().equals(point);
            }
        }));
    }

    private void hitTarget(Point point) {
        IObstacle obstacle = findObstacle(point);

        if (obstacle != null && obstacle.isDestructable()) {
            dstroyObstacle(obstacle);
        } else if (obstacle == null) {
            createObstacle(point);
        }
    }

    protected void dstroyObstacle(IObstacle obstacle) {
        mars.removeObstacle(obstacle);
    }

    protected void createObstacle(Point point) {
        IObstacle obstacle = new Crater(point);
        mars.addObstacle(obstacle);
    }
}

